pkg/parser/conversion.go

```diff
@@ -0,0 +1,41 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parser
```

Comment:
Should this package be in crossplane, or crossplane-runtime? It seems like there will be some consumers outside of crossplane/crossplane, right? They might prefer to take a dependency on crossplane-runtime.

Reply:
I generally look at `crossplane-runtime` as the place to put types and interfaces and ways to interact with those types and interfaces. Since the types being interacted with here only live in `crossplane/crossplane`, it feels like this package should be co-located. The _primary_ consumers of this package would be `crank` and the package controllers, so it would be nice for them to live close to each other. If someone was taking a dependency on this package they would likely be taking a dependency on `crossplane/crossplane` anyway to interact with the types.

Reply:
Would your concern still exist if this library operated on `runtime.Object` rather than specific types that were defined in crossplane/crossplane?

I take your point that this parser library is currently aware of types that are defined within this repo. I assumed that the primary consumers of this package would be:

* Crank.
* The package manager controllers.
* The entrypoint of our packages.
* The Upbound Cloud registry.

The former two seem like they'll live within in this repo. I'm less confident about the latter two. In general what I am trying to avoid here is crossplane/crossplane becoming both packaged tools (controller managers, crank) and a library that other projects take dependencies on. I'd like us to eventually be able to put everything in crossplane/crossplane inside an `internal` package such that it _cannot_ be imported by anyone else. This allows us to control which parts of our codebase are libraries that other projects or repos might eventually depend on - i.e. presumably the packages in crossplane-runtime.

Reply:
If Upbound Cloud registry were to use this parser then it's very likely that they'll need to include those types anyway for doing any kind of meaningful operation besides metadata ops. If we had a repo like `crossplane/api` with all core Crossplane types (maybe discuss that in an issue?), then having this in `crossplane-runtime` would've made more sense but right now, I highly doubt that moving this to `crossplane-runtime` will save any of the consumers from depending on `crossplane/crossplane` and this implementation will still bear the cost of working with `runtime.Object`.

Reply:
@muvaf I'll open an issue to track. In the meantime, it seems like putting this in `crossplane-runtime` is safer in the case that someone wants to consume it and does not need Crossplane types.

Reply:
I'm wary of `crossplane/crossplane` as an artifact being becoming many things; a controller manager, a CLI tool, and a set of libraries that folks import to make clients. I take your point that folks will need to import `crossplane/crossplane` anyhow for the API types, but I would prefer to address this by considering decoupling the API types. I'll comment in more detail on https://github.com/crossplane/crossplane/issues/1708 that @hasheddan raised about this.